src/crimson/osd/osd_operations/client_request.cc

```diff
@@ -215,15 +215,79 @@ auto ClientRequest::reply_op_error(const Ref<PG>& pg, int err)
   return get_foreign_connection().send_with_throttling(std::move(reply));
 }
 
+ClientRequest::interruptible_future<>
+ClientRequest::recover_missing_snaps(
+  Ref<PG> pg,
+  instance_handle_t &ihref,
+  std::set<snapid_t> &snaps)
+{
+  return pg->obc_loader.with_obc<RWState::RWREAD>(
+    m->get_hobj().get_head(),
+    [pg, this, &ihref, &snaps](auto head, auto) {
+    return ihref.enter_stage<interruptor>(
+      client_pp(*pg).recover_missing_snaps, *this
+    ).then_interruptible([pg, head, &snaps] {
+      return InterruptibleOperation::interruptor::do_for_each(
+        snaps,
+        [pg, head](auto &snap)
+        -> InterruptibleOperation::template interruptible_future<> {
+        auto coid = head->obs.oi.soid;
+        coid.snap = snap;
+        auto oid = resolve_oid(head->get_head_ss(), coid);
+        /* Rollback targets may legitimately not exist if, for instance,
+         * the object is an rbd block which happened to be sparse and
+         * therefore non-existent at the time of the specified snapshot.
+         * In such a case, rollback will simply delete the object.  Here,
+         * we skip the oid as there is no corresponding clone to recover.
+         * See https://tracker.ceph.com/issues/63821 */
+        if (oid) {
+          return do_recover_missing(pg, *oid);
+        } else {
+          return seastar::now();
+        }
+      });
+    });
+  }).handle_error_interruptible(
+    crimson::ct_error::assert_all("unexpected error")
+  );
+}
+
 ClientRequest::interruptible_future<>
 ClientRequest::process_op(
   instance_handle_t &ihref, Ref<PG> &pg, unsigned this_instance_id)
 {
   LOG_PREFIX(ClientRequest::process_op);
   return ihref.enter_stage<interruptor>(
     client_pp(*pg).recover_missing, *this
-  ).then_interruptible([pg, this]() mutable {
-    return recover_missings(pg, m->get_hobj(), snaps_need_to_recover());
+  ).then_interruptible([pg, this, FNAME] {
+    const hobject_t &soid = m->get_hobj();
+    if (!pg->is_primary()) {
+      DEBUGDPP(
+        "Skipping recover_missings on non primary pg for soid {}", *pg, soid);
+      return interruptor::now();
+    }
+    return do_recover_missing(pg, m->get_hobj().get_head());
+  }).then_interruptible([pg, this, &ihref] {
+    std::set<snapid_t> snaps = snaps_need_to_recover();
+    if (snaps.empty()) {
+      return interruptor::now();
+    }
+    return seastar::do_with(
+      std::move(snaps),
+      [pg, this, &ihref](auto &snaps) {
+      // call recover_missing_snaps() in order, but wait concurrently for loading.
+      auto maybe_fut = ihref.enter_stage_maybe_sync(
+          client_pp(*pg).recover_missing_lock_obc, *this);
+      if (maybe_fut.has_value()) {
+        return interruptor::make_interruptible(
+          std::move(maybe_fut.value())
+        ).then_interruptible([pg, this, &ihref, &snaps] {
+          return recover_missing_snaps(pg, ihref, snaps);
+        });
+      } else {
+        return recover_missing_snaps(pg, ihref, snaps);
```

Comment:
It seems like this std::optional<seastar::future> return from enter_stage_maybe_sync is to avoid a .then() on a ready future.  Is that actually worth the extra complexity?

Reply:
> It seems like this std::optional<future> return from enter_stage_maybe_sync is to avoid a .then() on a ready future. Is that actually worth the extra complexity?

Please refer to a similar change https://github.com/ceph/ceph/commit/c7628de4d230cd08d536d7e0c35bf6822cc7146e

The goal is to avoid `.then()` upon leaving an exclusive and entering a concurrent phase, to preserve the correct calling order of `recover_missing_snaps()` or `process_lock_obc()`, specifically, the calling order of `pg->obc_loader.with_obc()` or `pg->with_locked_obc()`.

The obc lock might be exclusive for the same object, and once locked, it will be held across multiple phases, so I suppose if the locking order of the same object is reordered, we will encounter dead-lock.

Reply:
Hm, probably I should instead check `ceph_assert(!maybe_fut.has_value())` here and with proper comment.

Reply:
So the caller must ensure that with_obc() (through recover_missing_snaps) is called synchronously following the call to ihref.enter_stage_maybe_sync, but only if std::nullopt is returned?

Reply:
I suppose in case of `with_obc()`, we should expect that the return value is always `std::nullopt` because `ihref.enter_stage_maybe_sync` is supposed to leave an exclusive phase (won't yield) and enter a concurrent phase (also won't yield) at the same time.

This is why I think the assert can be added https://github.com/ceph/ceph/pull/55488#discussion_r1506903525

Reply:
For this call site, we are transitioning from recover_missing (OrderedExclusive) to recover_missing_lock_obc (OrderedConcurrent).  We can be certain that transition will occur without blocking because exiting an OrderedExclusivePhase never blocks and entering an OrderedConcurrent phase never blocks.  However, there isn't enough type information in PipelineHandle to prove the above statically -- PipelineHandle::barrier could be populated.

The next call site

```
           // call process_lock_obc() in order, but wait concurrently for loading.
            auto maybe_fut = ihref.enter_stage_maybe_sync(
                client_pp(*pg).lock_obc, *this);
```

is a transition from get_obc (OrderedExclusive) to lock_obc (OrderedConcurrent).  Same story -- we know that this call will always return std::nullopt, but the type information in PipelineHandle is insufficient to prove it.

The final call site is in internal_client_request:

```
            // call process_lock_obc() in order, but wait concurrently for loading.
            auto maybe_fut = enter_stage_maybe_sync(client_pp().lock_obc);
            if (maybe_fut.has_value()) {
              return interruptor::make_interruptible(
                std::move(maybe_fut.value())
              ).then_interruptible([this, &osd_ops] {
                return process_lock_obc(osd_ops);
              });
```

This one is again a transition from get_obc (OrderedExclusive) to lock_obc (OrderedConcurrent) and will never block.

It seems to me that:
1. In all three cases, as currently written, enter_stage_maybe_sync will *always* return std::nullopt
2. If 1. ever changes, it will introduce a subtle bug as the transition to the second stage won't be synchronous with invoking with_obc()

So, I suggest that we replace enter_stage_maybe_block with enter_stage_sync.  enter_stage_sync returns void, statically asserts !T::is_enter_async, and dynamically asserts that barrier->wait() returns std::nullopt.

I am still fuzzy on the details here, so I may well be missing something, though.

Reply:
Yeah, reasonable to me.

Reply:
see https://github.com/ceph/ceph/pull/55488/commits/61ddc7499aefcf47a3e5167290880be902d936e2